Hooks/Chat.cs

```diff
@@ -21,53 +18,51 @@ public static class Chat
     /// </summary>
     public static event ChatEventHandler? OnChatMessage;
 
-    private static FastNativeDetour? Detour;
+    //private static INativeDetour? Detour;
+    private static Harmony? _harmony;
 
     public static unsafe void Initialize()
     {
-        if (Detour != null)
+        if (_harmony != null)
             throw new Exception("Detour already initialized. You don't need to call this. The Wetstone plugin will do it for you.");
 
-        var ty = typeof(ChatMessageSystem).GetNestedTypes().First(x => x.Name.Contains("ChatMessageJob"));
-        Detour = NativeHookUtil.Detour(ty, "OriginalLambdaBody", Hook, out Original);
+        _harmony = Harmony.CreateAndPatchAll(typeof(Chat));
     }
 
     public static unsafe void Uninitialize()
     {
-        if (Detour == null)
+        if (_harmony == null)
             throw new Exception("Detour wasn't initialized. Are you trying to unload Wetstone twice?");
 
-        Detour.Dispose();
-        Detour = null;
-
-        OnChatMessage = null;
+        _harmony.UnpatchSelf();
     }
 
-    private static unsafe void Hook(IntPtr _this, Entity* chatMessageEntity, ChatMessageEvent* chatMessage, FromCharacter* fromCharacter)
+    [HarmonyPrefix]
+    [HarmonyPatch(typeof(ChatMessageSystem), nameof(ChatMessageSystem.OnUpdate))]
+    public static void OnUpdatePrefix(ChatMessageSystem __instance)
     {
-        var ev = new VChatEvent(fromCharacter->User, fromCharacter->Character, chatMessage->MessageText.ToString(), chatMessage->MessageType);
-
-        WetstonePlugin.Logger.LogInfo($"[Chat] [{ev.Type}] {ev.User.CharacterName}: {ev.Message}");
-
-        try
+        var entities = __instance._ChatMessageQuery.ToEntityArray(Unity.Collections.Allocator.Temp);
+        foreach (var entity in entities)
         {
-            OnChatMessage?.Invoke(ev);
-
-            if (ev.Cancelled) return;
+            var chatMessage = VWorld.Server.EntityManager.GetComponentData<ChatMessageEvent>(entity);
+            var fromCharacter = VWorld.Server.EntityManager.GetComponentData<FromCharacter>(entity);
+            var ev = new VChatEvent(fromCharacter.User, fromCharacter.Character, chatMessage.MessageText.ToString(), chatMessage.MessageType);
+
+            WetstonePlugin.Logger.LogInfo($"[Chat] [{ev.Type}] {ev.User.CharacterName}: {ev.Message}");
+
+            try
+            {
+                OnChatMessage?.Invoke(ev);
+
+                if (ev.Cancelled) return;
```

Comment:
Does this work to cancel messages as intended? I think there may be something lost in the translation from the native version. If a downstream plugin marks the event as canceled the OnUpdate prefix exits processing the loop of entities, but the original chat message entity and it's processing seem unaffected. Also the query tends to have a single message, but in theory there could be more and then I think this just ignores those for a frame and then they could be unprocessed and exited by the next one. 

I'd not return here, then either 'not support' Cancelled anymore or simulate it for the callers.

You could considering using this whisper back from the VCF gloomrot changes to simulate canceling the message to the rest of the server: https://github.com/decaprime/VampireCommandFramework/commit/6871f4b00e3df1094856ff60817ed52b84723405#diff-4fca44e8b9150400cd4f27773e9541b0dfc90f2c50fab1819b960b5a7be6f896 .

Reply:
Hmm, i just moved the code to a new funciton. But the now that you mention it, the return doesn't make sense here.